### PR TITLE
implement caching of the dependency graph 

### PR DESCRIPTION
**Pull request by matthid**
to improve the speed of the integration tests suite.

If I can make this work for the test suite we might use it later for a global cache?

### CODE REVIEW

File: src/Paket.Core/Installation/UpdateProcess.fs [hunk 1]
```diff
@@ -140,7 +140,13 @@ let selectiveUpdate force getSha1 getSortedVersionsF getPackageDetailsF getRunti
...
         getVersionsF,getPackageDetailsF,groups
 
-    let resolution = dependenciesFile.Resolve(force, getSha1, getVersionsF, getPackageDetailsF, getRuntimeGraphFromPackage, groupsToUpdate, updateMode)
+    let resolution =
+        dependenciesFile.Resolve(
+            force, getSha1,
+            getVersionsF |> GraphCache.liftGetVersionsF,
+            getPackageDetailsF |> GraphCache.liftGetPackageDetailsF,
+            getRuntimeGraphFromPackage |> GraphCache.liftGetRuntimeGraphFromPackage,
+            groupsToUpdate, updateMode)
```
Comment (matthid) [inline src/Paket.Core/Installation/UpdateProcess.fs]: This is how I hook the cache into the code

File: src/Paket.Core/PaketConfigFiles/GraphCache.fs [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+// - We need to be carefull when caching the list of versions...
+let private tree = ConcurrentDictionary<NuGetSourceUri * PackageName, CachedInfo>()
+let updateCacheFile () =
+    tree
+    |> Seq.toList
+    |> Seq.map (fun kv ->
+        let uri, name = kv.Key
+        uri, name,kv.Value)
+    |> createGraphJson
+    |> fun j -> System.IO.File.WriteAllText(graphCacheFile, j.ToString())
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/GraphCache.fs]: Updating the file on every change is probably quite slow...

File: src/Paket.Core/PaketConfigFiles/GraphCache.fs [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+
+let readCacheFile () =
+    if System.IO.File.Exists graphCacheFile then
+        System.IO.File.ReadAllText(graphCacheFile)
+        |> JObject.Parse
+        |> readGraphJson
+        |> Seq.iter (fun (source, packName, cache) ->
+            tree.AddOrUpdate((source,packName), cache, (fun _ oldCache ->
+                // TODO merge...
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/GraphCache.fs]: ...

File: src/Paket.Core/PaketConfigFiles/GraphCache.fs [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+                    | None -> packageDetails, runtimeGraph) }
+
+    match tree.TryGetValue((source.Url,name)) with
+    | true, cache ->
+        Some (fixSource cache)
+    | _ -> None
+
+do
+    if isGraphCacheEnabled then
+        readCacheFile()
```
Comment (matthid) [inline src/Paket.Core/PaketConfigFiles/GraphCache.fs]: Global mutable state, yay

**Comment by matthid** [comment]
Seems like this yield the kind of performance I was expecting. @cloudRoutine is probably right that the cost of spawning new processes is what makes it slow? Or it is the download/extracting of the packages itself...

**Comment by forki** [comment]
+222,468 −12 

????

**Comment by matthid** [comment]
Its adding a cache for all the tests :)

**Comment by forki** [comment]
what does that mean? why do we commit that`?

**Comment by matthid** [comment]
how else would we have it on the build server ?

**Comment by forki** [comment]
Ok tbh I don't yet understand what we want to achieve.

**Comment by matthid** [comment]
increase the speed of the integration test suite

**Comment by forki** [comment]
It changes the full nature of the integration tests. Now we don't test
integration with nuget.org anymore. At least not on the same level.

Am 02.05.2017 1:45 nachm. schrieb "Matthias Dittrich" <
notifications@github.com>:

**Comment by cloudRoutine** [comment]
We waste a lot of time downloading and installing packages for tests whose primary function is not to test downloading and installing packages.

**Comment by forki** [comment]
It's not really a waste and who cares if it runs on CI?

**Comment by cloudRoutine** [comment]
The time matters when you're working on a feature, the package resolver in particular,  that's only thoroughly tested by the integration tests.

If they're not "integration" tests anymore then we can change them to unit tests then. 

Installation has nothing to with "nuget integration" and it is a waste of time on tests that don't do anything with packages after they've been resolved or installed.

**Comment by 0x53A** [comment]
Another issue with committing the graph is that we don't catch issues related to updated package versions.

It would probably be useful to test both - that the graph how it was when the test was created still resolves, and that the graph also resolves with the current package versions.

**Comment by matthid** [comment]
Yes and additionally it helps to find regressions in the resolver, because if the resolver decides it needs additional info from nuget after some changes we can make that red. I think the pinning timeout problem could have been caught with this

**Comment by forki** [comment]
tbh I think runtime is an issue here. Especially for the resolver development you want to pin edge cases in unit tests. I think integration tests should be as real world as possible and some should use caches but not all.

**Comment by matthid** [comment]
Yes that's the plan, this is very early playing around... Obviously I enabled for all atm to see impact and test the impl itself

**Comment by matthid** [comment]
Note as I said above I'm not happy with the results. so it's possible that I close this as failed attempt if there are no other use cases this makes sense

**Comment by matthid** [comment]
I definitely shows that retrieving metadata is not our bottleneck ;)

**Comment by matthid** [comment]
Not sure if it was the right approach, it didn't yield any interesting results
